include/envoy/network/io_handle.h

```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/common/pure.h"
+
+namespace Envoy {
+namespace Network {
+
+/**
+ * IoHandle: an abstract interface for all I/O operations
+ */
+class IoHandle {
+public:
+  IoHandle() {}
+
+  virtual ~IoHandle() {}
+
+  // TODO(sbelair2) remove fd() method
+  virtual int fd() const PURE;
+
+  /**
+   * @param the socket file descriptor to set in the handle. Assigns an fd from
+   * an external socket operation such as from libevent or the dispatcher after construction
+   *
+   * TODO(sbelair2):  To be removed when the IoSocketHandle derivative is integrated
+   * and the fd is fully abstracted from clients.
+   */
+  virtual void operator=(int fd) PURE;
```

Comment:
Do we need this at the interface level? In general I prefer to avoid this type of stuff if possible. Can this just be handled via construction of a concrete sub-class?

Reply:
I understand. Unfortunately, it seems that we need this to set the fd in an already-constructed fd. So we either have a method to set the fd, or we have the assignment operator. Is there another way to do this?

Reply:
Is there a usage where the FD is set to something other than -1?

Reply:
This is exactly where it is needed. As you can see, there's already a TODO here. I'm inclined to leave the code as is, with the TODO, ok?

Reply:
Per @jmarantz can you just add a method like `close()` or whatever is the purpose of setting it to -1? I think that would be much clearer and less easy to accidentally abuse.

Reply:
Yes, will do. @jmarantz thanks for the suggestion- much better way to do it.

Reply:
@mattklein123 It turns out that the assignment operator is needed for at least one test (OriginalDstTestFilter). Please advise.

Reply:
Look at the test, figure out what it's doing, and fix in some sensible way? It makes no sense to assign -1 to this object so just figure out the semantics and come up with an object oriented solution?